src/producer/__tests__/idempotentProduceMessages.spec.js

```diff
@@ -0,0 +1,222 @@
+    await Promise.all([producer.connect(), consumer.connect()])
+    await consumer.subscribe({ topic: topicName, fromBeginning: true })
+  })
+
+  afterEach(
+    async () =>
+      await Promise.all([
+        producer && (await producer.disconnect()),
+        consumer && (await consumer.disconnect()),
+      ])
+  )
+
+  it('sequential produce() calls > all messages are written to the partition once, in order', async () => {
+    const messagesConsumed = []
+
+    for (const m of messages) {
+      await producer.send({ acks: -1, topic: topicName, messages: [m] })
+    }
+
+    await consumer.run({ eachMessage: async message => messagesConsumed.push(message) })
+
+    await waitForMessages(messagesConsumed, { number: messages.length })
+
+    messagesConsumed.forEach(({ message: { value } }, i) =>
+      expect(value.toString()).toEqual(`${i}`)
+    )
+  })
+
+  it('sequential produce() calls > where produce() throws a retriable error, all messages are written to the partition once, in order', async () => {
+    for (const nodeId of [0, 1, 2]) {
+      const broker = await cluster.findBroker({ nodeId })
+
+      const brokerProduce = jest.spyOn(broker, 'produce')
+      brokerProduce.mockImplementationOnce(async () => {
+        await sleep(5)
+        throw new KafkaJSError('retriable error')
+      })
+    }
+
+    const messagesConsumed = []
+
+    for (const m of messages) {
+      await producer.send({ acks: -1, topic: topicName, messages: [m] })
+    }
+
+    await consumer.run({ eachMessage: async message => messagesConsumed.push(message) })
+
+    await waitForMessages(messagesConsumed, { number: messages.length })
+
+    messagesConsumed.forEach(({ message: { value } }, i) =>
+      expect(value.toString()).toEqual(`${i}`)
+    )
+  })
+
+  it('sequential produce() calls > where produce() throws a retriable error after the message is written to the log, all messages are written to the partition once, in order', async () => {
+    for (const nodeId of [0, 1, 2]) {
+      const broker = await cluster.findBroker({ nodeId })
+      const originalCall = broker.produce.bind(broker)
+      const brokerProduce = jest.spyOn(broker, 'produce')
+      brokerProduce.mockImplementationOnce()
+      brokerProduce.mockImplementationOnce()
+      brokerProduce.mockImplementationOnce(async (...args) => {
+        await originalCall(...args)
+        throw new KafkaJSError('retriable error')
+      })
+    }
+
+    const messagesConsumed = []
+
+    for (const m of messages) {
+      await producer.send({ acks: -1, topic: topicName, messages: [m] })
+    }
+
+    await consumer.run({ eachMessage: async message => messagesConsumed.push(message) })
+
+    await waitForMessages(messagesConsumed, { number: messages.length })
+
+    messagesConsumed.forEach(({ message: { value } }, i) =>
+      expect(value.toString()).toEqual(`${i}`)
+    )
+  })
+
+  it('concurrent produce() calls > all messages are written to the partition once', async () => {
```

Comment:
Don't worry about changing this, but just for future reference, you can just wrap these test cases in a describe block to group the tests together, instead of repeating the prefix in each test case:

```js
describe('concurrent produce() calls', () => {
  test('all messages are written to the partition at once')
})
```